[Flang] LoongArch64 support for BIND(C) derived types in mabi=lp64d.

ylzsx opened the pull request:
This patch:
- Supports both the passing and returning of BIND(C) type parameters.
- Adds `mabi` check for LoongArch64. Currently, flang only supports `mabi=` option
  set to `lp64d` in LoongArch64, other ABIs will report an error and may be supported 
  in the future.

Reference ABI:
https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#subroutine-calling-sequence

ylzsx committed:
[Flang] LoongArch64 support for BIND(C) derived types.

This patch supports both the passing and returning of BIND(C) type parameters.

Reference ABI:
https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#subroutine-calling-sequence

SixWeining commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1081,6 +1081,9 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
   using GenericTarget::GenericTarget;
 
   static constexpr int defaultWidth = 64;
+  static constexpr int GRLen = defaultWidth; /* eight bytes */
```
I wonder how flang dealing with different ABIs?

ylzsx replied:
Thanks for your review. This patch only implements `mabi=lp64d` in LoongArch64, and does not currently support others,  though they may be supported in the future. To make it more explicit, I will modify this patch to report an error when other ABIs or conflicting options are detected.

ylzsx committed:
[Flang] Only `lp64d` abi is supported in LoongArch64.

This patch:
- Adds `mabi` check for LoongArch64. Currently, flang only supports `mabi=` option
set to `lp64d` in LoongArch64, other ABIs will report an error and may be supported
in the future.

SixWeining commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -807,6 +807,10 @@ def err_drv_loongarch_invalid_simd_option_combination : Error<
   "invalid argument '%0' to -msimd=; must be one of: none, lsx, lasx">;
+// Support for other float specifications for flang in LoongArch64 may be
+// added in the future, such as 32-bit fpu or soft fpu.
+def err_drv_loongarch_disable_fpu64_for_flang : Error<
```
Can we use `err_drv_argument_not_allowed_with` instead?

SixWeining commented on flang/test/Driver/mabi-loongarch.f90:
```diff
@@ -0,0 +1,13 @@
+! RUN: not %flang -### -c --target=loongarch64-unknown-linux -mabi=lp64s %s 2>&1 | FileCheck --check-prefix=INVALID1 %s
```
Usually `###` is the last option.
```suggestion
! RUN: not %flang -c --target=loongarch64-unknown-linux -mabi=lp64s %s ### 2>&1 | FileCheck --check-prefix=INVALID1 %s
```

SixWeining commented on flang/test/Driver/mabi-loongarch.f90:
```diff
@@ -0,0 +1,13 @@
+! RUN: %flang -### -c --target=loongarch64-unknown-linux -mabi=lp64d %s 2>&1 | FileCheck --check-prefix=ABI %s
+! RUN: %flang -### -c --target=loongarch64-unknown-linux %s 2>&1 | FileCheck --check-prefix=ABI %s
+
+! REQUIRES: target=loongarch64{{.*}}
```
Unnecessary requirement.

ylzsx committed:
Fixes for review

ylzsx commented:
Hi, @tblah, Could you take a look and provide your feedback? Thanks in advance! It supports both the passing and returning of BIND(C) type parameters for LoongArch.

tblah commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,311 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+                mlir::IntegerType::get(type.getContext(), width));
+        })
+        .template Case<fir::CharacterType>([&](fir::CharacterType charTy) {
+          flatTypes.push_back(mlir::IntegerType::get(type.getContext(), 8));
```
Character types can have different kinds in the same way as logical types. For example,
```
character(kind=4) :: c
```
In this case `c` would be 4 bytes long.

tblah replied:
Okay that sounds plausible. Please could you add an assertion that the character has size 1 (mostly to document the assumption). Other than that the current approach looks good.

ylzsx replied:
Ok, I will add an assertion to document the assumption and address the other feedback as well. Thank you very much for your valuable suggestions!

tblah commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,311 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+  /// when flattened it contains a single fp value, fp+fp, or int+fp of
+  /// appropriate size).
+  bool detectFARsEligibleStruct(mlir::Location loc, fir::RecordType recTy,
+                                mlir::Type &Field1Ty,
```
nit: This part of flang uses the MLIR programming style: https://mlir.llvm.org/getting_started/DeveloperGuide/#style-guide so local variables should be named with the first letter as lower case. In this case `field1Ty`.

tblah commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,311 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+    assert((mlir::isa<mlir::IntegerType, mlir::FloatType>(flatTypes[0])) &&
+           "Type must be int or float after flattening");
+    if (auto floatTy = mlir::dyn_cast<mlir::FloatType>(flatTypes[0])) {
+      auto Size = floatTy.getWidth();
```
nit: please spell out `auto` in cases where the type is not spelled out in the statement. For example,

`auto floatTy = mlir::dyn_cast<mlir::FloatType>(flatTypes[0])` is okay because `mlir::FloatType` is clear from reading the line.

`auto size = floatTy.getWidth();` is not okay because one can't see what the type is here.

More info here https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

tblah commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,311 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+    llvm::TypeSwitch<mlir::Type>(type)
+        .template Case<mlir::IntegerType>([&](mlir::IntegerType intTy) {
+          const auto width = intTy.getWidth();
+          assert(width <= 128 &&
+                 "integer type with width more than 128 bits is unexpected");
```
nit: I don't think flang supports integers larger than 128 bits, but it would be easy to forget to update this line if that were added later. I think this should be a `TODO()` so that non-assertion builds do not generate incorrect code in this case.

ylzsx committed:
Fixes for review

tblah reviewed (approved):
The code changes look good to me. I am not at all familiar with LoongArch64 so it would be best if somebody else reviews that this implements the ABI correctly.

SixWeining commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,317 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+                          const Marshalling &previousArguments,
+                          const mlir::Type &field1Ty,
+                          const mlir::Type &field2Ty) const {
+
```
useless blank line

SixWeining commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,317 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+
+  bool checkTypehasEnoughReg(mlir::Location loc, int &GARsLeft, int &FARsLeft,
+                             const mlir::Type type) const {
+    if (type == nullptr)
```
Is this a legal comparison?

ylzsx committed:
Fixes for review: VectorType ABI handle and others.

Currently, the function `fir::getTypeSizeAndAlignmentOrCrash` does
not yet handle the size and alignment for `fir::VectorType`, but we
are still using it for now. As a result, it will report a TODO message,
and this functionality will be implemented in a future patch.

ylzsx commented on flang/lib/Optimizer/CodeGen/Target.cpp:
```diff
@@ -1151,6 +1154,317 @@ struct TargetLoongArch64 : public GenericTarget<TargetLoongArch64> {
+              llvm::copy(subTypeList, std::back_inserter(flatTypes));
+          }
+        })
+        .template Case<fir::VectorType>([&](fir::VectorType vecTy) {
```
Thanks for your review. This is my oversight. According to the [LoongArch ABI](https://github.com/loongson/la-abi-specs/blob/release/lapcs.adoc#vectors-1), 128-bit vector are passed with two GARs with adjacent numbers (if available), and 256-bit vector are passed on the stack. In other cases, the manual has not explicitly specified. 

I will modify my code to comply with this ABI requirement and report a TODO for other cases. Specifically,
* For a vector smaller than 128 bits, report a TODO.
* For a vector equal to 128 bits, return `i128`
* For a vector larger than 128 bits, pass it on the stack.

I will try to make sure it behaves as similar to clang as possible. Looking forward to your suggestions.

ylzsx committed:
Merge branch 'main' into binc-pr

SixWeining reviewed (approved):
LGTM for the LoongArch bits.
